File: javascript/dnd-character/dnd-character.js
//
// This is only a SKELETON file for the 'D&D Character' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const abilityModifier = (score) => {
  if (score < 3) {
    throw new Error('Ability scores must be at least 3');
  } else if (score > 18) {
    throw new Error('Ability scores can be at most 18');
  };
  return Math.floor((score - 10) / 2);
};

export class Character {
  static rollAbility() {
    let throws = [];
    for (let i = 0; i < 4; i++) {
      throws.push(Math.floor(Math.random() * 6 + 1));
    }
    throws.sort();
    return throws[1] + throws[2] + throws[3];
  }

  get strength() {
    if (this.strengthScore === undefined) {
      this.strengthScore = Character.rollAbility()
    }
    return this.strengthScore;
  }

  get dexterity() {
    if (this.dexterityScore === undefined) {
      this.dexterityScore = Character.rollAbility()
    }
    return this.dexterityScore;
  }

  get constitution() {
    if (this.constitutionScore === undefined) {
      this.constitutionScore = Character.rollAbility()
    }
    return this.constitutionScore;
  }

  get intelligence() {
    if (this.intelligenceScore === undefined) {
      this.intelligenceScore = Character.rollAbility()
    }
    return this.intelligenceScore;
  }

  get wisdom() {
    if (this.wisdomScore === undefined) {
      this.wisdomScore = Character.rollAbility()
    }
    return this.wisdomScore;
  }

  get charisma() {
    if (this.charismaScore === undefined) {
      this.charismaScore = Character.rollAbility()
    }
    return this.charismaScore;
  }

  get hitpoints() {
    return 10 + abilityModifier(this.constitutionScore);
  }
}

File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.List;
import java.util.ArrayList;

class DnDCharacter {

    private int strength;
    private int dexterity;
    private int constitution;
    private int intelligence;
    private int wisdom;
    private int charisma;
    private int hitpoints;

    public DnDCharacter() {
        this.strength = ability(rollDice());
        this.dexterity = ability(rollDice());
        this.constitution = ability(rollDice());
        this.intelligence = ability(rollDice());
        this.wisdom = ability(rollDice());
        this.charisma = ability(rollDice());
        this.hitpoints = modifier(this.constitution) + 10;
    }

    int ability(List<Integer> scores) {
        int min = scores.get(0);
        int total = 0;
        for (int i = 1; i <= 3; i++) {
            int value = scores.get(i);
            if (value < min) {
                total += min;
                min = value;
            } else {
                total += value;
            }
        }
        return total;
    }

    List<Integer> rollDice() {
        List<Integer> dice = new ArrayList<>();
        for (int i = 0; i < 4; i++) {
            dice.add((int) Math.floor(Math.random() * 6 + 1));
        }
        return dice;
    }

    int modifier(int input) {
        return input / 2 - 5;
    }

    int getStrength() {
        return this.strength;
    }

    int getDexterity() {
        return this.dexterity;
    }

    int getConstitution() {
        return this.constitution;
    }

    int getIntelligence() {
        return this.intelligence;
    }

    int getWisdom() {
        return this.wisdom;
    }

    int getCharisma() {
        return this.charisma;
    }

    int getHitpoints() {
        return this.hitpoints;
    }
}
